lib/measure-text.js

```diff
@@ -1,31 +1,114 @@
+    return this.document.widthOfString(str);
+  }
+
+  static create(fontPath, fallbackFontPath) {
+    return new this().useFont(fontPath)
+      .catch(err => {
+        if (fallbackFontPath) {
+          console.log(`Unable to load font at ${fontPath}. Trying fallback font ${fallbackFontPath}.`);
+          return new this().useFont(fallbackFontPath);
+        } else {
+          throw err;
+        }
+      });
   }
 }
 
-loadFont(path.join(__dirname, '..', 'Verdana.ttf'), function (err) {
-  if (err && process.env.FALLBACK_FONT_PATH) {
-    loadFont(process.env.FALLBACK_FONT_PATH);
+class QuickTextMeasurer {
+  constructor(baseMeasurer) {
+    Object.assign(this, { baseMeasurer });
+
+    // This will be a Map of characters -> numbers.
+    this.characterWidths = new Map();
+    // This will be Map of Maps of characters -> numbers.
+    this.kerningPairs = new Map();
   }
-});
-doc = doc.fontSize(11);
 
-function measure(str) {
-  return doc.widthOfString(str);
+  static create(fontPath, fallbackFontPath) {
+    return PDFKitTextMeasurer.create(fontPath, fallbackFontPath)
+      .then(pdfKitMeasurer => {
+        const result = new this(pdfKitMeasurer);
+        result.prepare();
+        return result;
+      });
+  }
+
+  static printableAsciiCharacters() {
+    const printableRange = [32, 126];
+    const length = printableRange[1] - printableRange[0] + 1;
+    return Array
+      .from({ length }, (value, i) => printableRange[0] + i)
+      .map(charCode => String.fromCharCode(charCode));
+  }
+
+  prepare() {
+    const charactersToCache = this.constructor.printableAsciiCharacters();
+
+    charactersToCache.forEach(char => {
+      this.characterWidths.set(char, this.baseMeasurer.widthOf(char));
+      this.kerningPairs.set(char, new Map());
+    });
+
+    charactersToCache.forEach(first => {
+      charactersToCache.forEach(second => {
+        const individually = this.characterWidths.get(first) + this.characterWidths.get(second);
+        const asPair = this.baseMeasurer.widthOf(`${first}${second}`);
+        const kerningAdjustment = asPair - individually;
+        this.kerningPairs.get(first).set(second, kerningAdjustment);
+      });
+    });
+  }
+
+  widthOf(str) {
+    const { characterWidths, kerningPairs } = this;
+
+    let result = 0;
+    let previous = null;
+    for (const character of str) {
+      if (!characterWidths.has(character)) {
+        // Bail.
+        return this.baseMeasurer.widthOf(str);
+      }
+
+      result += characterWidths.get(character);
+      if (previous !== null) {
+        result += kerningPairs.get(previous).get(character);
+      }
+
+      previous = character;
+    }
+    return result;
+  }
 }
 
-module.exports = measure;
-module.exports.loadFont = loadFont;
+// const defaultMeasurer = new TextMeasurer();
+// whenReady.then(() => defaultMeasurer.prepare());
+
+module.exports = {
+  PDFKitTextMeasurer,
+  QuickTextMeasurer,
+  // measure: defaultMeasurer.measure.bind(defaultMeasurer),
```

Comment:
> I liked having the measure function as the default export.

Will try to restore that. Was in the middle of cleaning up a gnarly module state bug, though if the initialization can all be done synchronously that will help a lot. Building the measurer is a little slow, so it would probably be good to avoid doing it when it's not necessary. Will think about a way to do that.

> Could we simply have initialization be synchronous?

Totally! I completely missed that. `loadFont` was using a callback before and I robotically ported it to promises, not realizing all the work was synchronous. That's great. It'll simplify this a lot.

Reply:
I ended up removing this, because the quick measurer takes a while to generate, and it was slowing down the CLI and the CLI tests. Plus it makes running exclusive tests longer, as the cache needs to be built even if it's not used by the tests being run.